### DISCUSSION TITLE
OSError 23

### DISCUSSION THREAD

**Discussion by dannybr** [micropython/micropython#14148] [Core Development]
OSError 23 is not listed in, i.e  module errno list does not include it.
anyways, from Unix it means Too many open files.

it happens to me, after restarting my ap some 5 times, and so far the only way to fix this, is
calling machine.reset(), but that brings me back to REPL,
so Q: 1- how can I close all open fd 's at the begining of my app?
or how to make sure my app starts in a clean environment?

**Comment by sophiedegran** [comment]
Hi
1- version of micropython ?
2- board or unix/ windows ?
3- did you open socket ?
...

**Comment by dannybr** [comment]
> On 21 Mar 2024, at 19:59, sophiedegran ***@***.***> wrote:
> 
> 
> Hi
> 1- version of micropython ?
> 
      MicroPython-1.22.2-riscv-IDFv5.0.4-with-newlib4.1.0
> 2- board or unix/ windows ?
> 
Board, tried on esp32 and esp32c3
> 3- did you open socket ?
> 
Yes, 2

**Reply by sophiedegran** [reply]
When you close your program did you close the socket ? 
I mean by default micropython can open 5 sockets max ...
And a socket still open after a closed program...  if you are in async mode you have to (example)
self server.close ()
await self.server.wait_closed ()
Hope that help it seem to be that

**Comment by dannybr** [comment]
The first issue is that the program never ends.
It binds to 2 sockets and waits for connection, and I’m using threads.
This works fine, the problem arises when I restart it, say after
some software change.
I assume that once it runs from the flash that won’t be a problem, but so far
Im avoiding that since every time I do that, it’s not easy to recover control (ie Thonny nor
Mpremote can stop it.  And once it’s running from flash, it’s not easy to diagnose
the problem :-)
Q: in C, I can restart the program with esp_restart(), but I don’t see an equivalent
In micropython.

**Reply by sophiedegran** [reply]
Yes I understand ... and why not use async mode with create_task ? It's easy to close..
Good luck

**Reply by andrewleech** [reply]
As suggested I'd personally use asyncio to manage multiple sockets, it's usually more reliable than threads.

If you want to reset the board from code, here's the docs: https://docs.micropython.org/en/latest/library/machine.html#reset-related-functions
